sycl/include/sycl/detail/generic_type_traits.hpp

```diff
@@ -493,13 +493,14 @@ using select_cl_scalar_integral_t =
 template <typename T>
 using select_cl_scalar_t = conditional_t<
     std::is_integral<T>::value, select_cl_scalar_integral_t<T>,
-    conditional_t<std::is_floating_point<T>::value, select_cl_scalar_float_t<T>,
-                  // half is a special case: it is implemented differently on
-                  // host and device and therefore, might lower to different
-                  // types
-                  conditional_t<std::is_same<T, half>::value,
-                                sycl::detail::half_impl::BIsRepresentationT,
-                                select_cl_scalar_complex_or_T_t<T>>>>;
+    conditional_t<
+        std::is_floating_point<T>::value, select_cl_scalar_float_t<T>,
+        // half is a special case: it is implemented differently on
+        // host and device and therefore, might lower to different
+        // types
+        conditional_t<std::is_same<std::remove_const_t<T>, half>::value,
```

Comment:
I think we can move volatile too, just for good measure:

```suggestion
        conditional_t<std::is_same<std::remove_cv_t<T>, half>::value,
```

Similar could be applied `type_helper` I believe.